User namespaces doc changes for 1.30 (#45178)
kubernetes/website: content/en/docs/concepts/workloads/pods/user-namespaces.md

Some popular filesystems that support idmap mounts in Linux 6.3 are: btrfs,
ext4, xfs, fat, tmpfs, overlayfs.

In addition, support is needed in the container runtime and in the OCI runtime
(the low level runtime used by the container runtime).

> The OCI runtime requirements are as follows:

Reviewer: ```suggestion
Here is a list of OCI runtimes known to provide support:
```
This is important. If someone adds the right support to another bit of code, this page shouldn't become inaccurate because of work done in another project. It's not fair to other projects to list a requirement based on stale information, and as a CNCF policy we avoid getting into that kind of situation.

If we'd more capacity, we could commit to keeping up to date with the ecosystem; we don't have that capacity, so we write in a way that means we don't need to stay current.

 * [crun][crun]: version 1.9 or greater (1.13+ recommended)

> Please note that [runc][runc] is the default OCI runtime in several Kubernetes
> distributions. Version `1.1.z` of runc doesn't support all the features needed
> by the Kubernetes implementation. If there is a newer release of runc than 1.1
> available for use, check its documentation and release notes for compatibility
> (look for idmap mounts support in particular, as that is the missing feature).

Reviewer: Try this wording:
```suggestion

<!-- ideally, update this if a newer minor release of runc comes out, whether or not it includes the idmap support -->
{{< note >}}
Many OCI runtimes do not include the support needed for using user namespaces in Linux pods.
If you use a managed Kubernetes, or have downloaded it from packages and set it up, it's likely
that nodes in your cluster use a runtime that doesn't include this support. For example, the most
widely used OCI runtime is `runc`, and version `1.1.z` of runc doesn't support all the features
needed by the Kubernetes implementation of user namespaces.
If there is a newer release of runc than 1.1 available for use, check its documentation and release
notes for compatibility (look for idmap mounts support in particular, because that is the missing feature).
{{< /note >}}
```

This aligns better with our style guide, and the comment highlights that it's intentionally not timeless (and that's fine; it's a style **guide** not a style rule set). We don't have to update this note every time that there are new runc releases, but we'd do well to maintain it as a best effort.

Author: Applied.

Some popular filesystems that support idmap mounts in Linux 6.3 are: btrfs,
ext4, xfs, fat, tmpfs, overlayfs.

In addition, support is needed in the container runtime and in the OCI container

> runtime.

Reviewer: Readers are likely to not know what you mean here. Can you clarify?

We might say “CRI-level runtime” and “OCI container runtime”, but then I think we'd do well to add an explanation about the two (maybe to our glossary?)

Author: Nah, that seems to complex and we are already using all over the place (not just here) the term "container runtime", I don't think we should change it.

I've adjusted to refer to the "OCI runtime" and added a clarification here too.

Reviewer: We've found a place where we need to explain two kinds of container runtimes. You and I know the difference, but plenty of readers will not.

If we don't explain it here, we should expect to be explaining it in GitHub issues (and I'd like to avoid that happening).

Author: @sftim What? I've changed it and explained, have you read what I changed? I don't think using "CRI runtime" is worth it, for the things I explained. It is just that.

Reviewer: Reviewing on behalf of SIG Docs, I'm disagreeing - I think we should help readers more. Can you convince me / us that we won't have confused readers here?

Author: @sftim sure. What do you think is not clear or could cause confusion? It would be great if you share that, to explain.

Let me explain in general, though, what has been already done:

First, I don't think we should use "CRI container runtime". We already use the term "container runtime" and we have been using it for several years, I don't see why we would change that. Furthermore, it is all over the place, it is not easy to change.

Secondly, the OCI runtime is only relevant for documentation in the userns context and maybe even for a few months only. So using a very specific term, instead of changing "container runtime" to "CRI container runtime" seems better.

Third, you asked to clarify and I chose a better name and clarified here what it is already.

What do you think is not clear and could cause confusion?

> The OCI container runtime requirements are as follows:
> 
>  * [crun][crun]: version 1.9 or greater (1.13+ recommended)
> 
> Please note that [runc][runc] is the default OCI container runtime in several
> Kubernetes distributions. Version `1.1.z` of runc doesn't support all the
> features needed by the Kubernetes implementation. If there is a newer release of
> runc than 1.1 available for use, check its documentation and release notes for
> compatibility (look for idmap mounts support in particular, as that is the
> missing feature).

Reviewer: Try this:

```suggestion

### Container runtime compatibility

To use user namespaces with Kubernetes, you need to use an OCI container runtime that
provides support for idmapped mounts; check the list below or verify with your chosen
runtime that it offers the features that Kubernetes needs.

{{% thirdparty-content single="true" %}}

Here is a list of OCI container runtimes that provide the features required for
user namespaces in Linux pods:

* [crun][crun]: version 1.9 or greater
  * Using version 1.13 or later provides the best overall experience

```

We very very rarely provide lists of software that isn't compatible (it could be a long list). There's an exception for containerd specifically because we previously did list it as compatible.

Reviewer: See https://github.com/kubernetes/website/pull/45178#issuecomment-1994316624

Author: I've seen it, I can't see why it is relevant. It says:

> I don't know of any place where Kubernetes hard codes runc as a kubelet default, or Pod API default, or RuntimeClass default; if is is, though, I can raise an issue (we should document this defaulting).
> 
>So far as I know, runc is very popular but Kubernetes is agnostic about which OCI runtime you use.

I didn't claim Kubernetes hardcodes anything about runc, so I really don't understand why you say that.

My point is: it is a widely used OCI runtime, therefore it is relevant information I want to provide to users.

Author: @sftim this text was changed several times, last time by @drewhagen suggestions. Do you still think this suggestion is better than what is currently written?

Reviewer: Let's mark this done and re-review

> The requirements for the 

Reviewer: ```suggestion
To use user namespaces with Kubernetes, you also need to use a CRI
```

Author: Didn't do the CRI change, so I didn't apply this.

Reviewer: I'm recommending the clarification around the two kind of container runtimes to help readers; for beta quality, we want to reduce the chance of a reader feeling confused (for alpha, we're less bothered).

Author: Can you read the changes I pushed and comment on that, if this is still not clear?

Reviewer: (sure, I can confirm I was reviewing based on the state of the PR at the time I commented; https://github.com/kubernetes/website/pull/45178/files#r1523158216 was considering the commits at that time)

Reviewer: This change: https://github.com/kubernetes/website/pull/45178#discussion_r1522133388

If it's a bad idea to make it and will harm readers' understanding, let's not do it. Otherwise, I'm asking for that change as a docs approver. Please consider accepting it.

Author: Applied.

Some popular filesystems that support idmap mounts in Linux 6.3 are: btrfs,
ext4, xfs, fat, tmpfs, overlayfs.

In addition, the container runtime and its underlying OCI runtime must support
user namespaces. The following OCI runtimes offer support:

> * [crun][crun] version 1.9 or greater (we recommend version 1.13+).

Reviewer: - (nit) We [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we
- (nit) It's easier for localizations to put the hyperlink URL inline

Author: Fixed, thanks

> Some popular filesystems that support idmap mounts in Linux 6.3 are: btrfs,

Reviewer 2: I'm curious, is there any way we can present the concept of idmap mounts for newcomers that aren't familiar with this from Linux?
https://lwn.net/Articles/896255/

Reviewer: How I'd cover that:
- have a blog article that explains Kubernetes and idmapped mounts
- publish that article a week or so before the release
- mark the article as `evergreen` (ie: we'll maintain it; usually we don't maintain blog articles once they reach a year old)
- link from these user namespaces docs to that evergreen blog article

An example of an evergreen article: https://kubernetes.io/blog/2020/09/03/warnings/

Author: IMHO that seems like too much overhead. It's not clear the value it would have to have such an ongoing investment to keep it up to date, as it is suggested. There are other places that do document idmap mounts, not sure I see the point of a k8s specific documentation.

Furthermore, Kubernetes users just need to know if a filesystem supports it or not.

ext4, xfs, fat, tmpfs, overlayfs.

In addition, support is needed in the container runtime and in the OCI runtime

> (the low level runtime used by the container runtime).

Reviewer 3: ```suggestion
(the lower level runtime used by the container runtime which satisfies the CRI).
```

Author: I've applied [this suggestion](https://github.com/kubernetes/website/pull/45178#discussion_r1525720052) and we don't have that parenthesis anymore. So I didn't apply this, but please re-open if you think we should clarify that in some other way :)

Here is a list of OCI runtimes known to provide support:

>  * [crun][crun]: version 1.9 or greater (1.13+ recommended)

Reviewer: The style guide also asks us to [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we) - and avoiding “we” is the more important tip here, I judge.

>  * [crun][crun]: version 1.9 or greater (1.13+ recommended)

Author: Applied, thanks.

## Set up a node to support user namespaces

By default, the kubelet assumes host's files and host's processes use UIDs/GIDs
in the range of 0-65535. This is the case in most Linux distributions with the
default installation. The kubelet will assign UIDs/GIDs higher than that to
pods, so there is no overlap between UID/GIDs used by host's files and processes
and the ones used by pods.

> Avoiding the overlap is important to mitigate the impact of CVEs like

Reviewer: Try this:
```suggestion
Avoiding the overlap is important to mitigate the impact of vulnerabilities such as
```
CVE is actually a database, and in docs it helps to be more precise than you would in speech or in a blog article

[CVE-2021-25741][CVE-2021-25741], where a pod can potentially read arbitrary
files in the host. If the UIDs/GIDs of the pod and the host don't overlap, it is
limited what a pod would be able to do: the pod UID/GID won't match the host's
file owner/group.

The kubelet can use a custom range for user IDs and group IDs for pods. To
configure a custom range, the node needs to have:

 * A user `kubelet` in the system (you cannot use any other username here)

>  * The binary `getsubids` installed (part of [shadow-utils][shadow-utils])

Reviewer: Let's tell the reader it needs to be in the `PATH` of the kubelet. I'm confident someone will manage to get that wrong if we're not specific; we have a _lot_ of users.

>  * The binary `getsubids` installed (part of [shadow-utils][shadow-utils])

Author: Fixed.

## Set up a node to support user namespaces

By default, the kubelet assumes host's files and host's processes use UIDs/GIDs
in the range of 0-65535. This is the case in most Linux distributions with the
default installation. The kubelet will assign UIDs/GIDs higher than that to
pods, so there is no overlap between UID/GIDs used by host's files and processes
and the ones used by pods.

> Note that this recommendation is important to mitigate the impact of CVEs like

Reviewer: ```suggestion
The assumption and default setting for user and group ID ranges is important to mitigate
the impact of security vulnerabilities such as
```

Author: I guess what you want to fix here is that we don't say "recommended" anymore now? I think it is clear as it is, but changing it to this instead: 

```
Avoiding the overlap is important to mitigate....
```

[CVE-2021-25741][CVE-2021-25741], where a pod can potentially read arbitrary
files in the host. If the UIDs/GIDs of the pod and the host don't overlap, it
is limited what a pod would be able to do: the pod UID/GID won't match the
host's file owner/group.

> The kubelet can use a custom range for pods too. To configure a custom range,

Reviewer: ```suggestion
The kubelet can use a custom ranges of user IDs and group IDs for pods. To configure a custom range,
```

the node needs to have:

>  * A user `kubelet` in the system

Reviewer: ```suggestion
 * A local POSIX user named `kubelet` (you cannot use any other username here)
```

Reviewer: I'd expect to find `kubelet` listed as a local user using `getent`. The first column in /etc/subuid is a login name that you can find using [NSS](https://en.wikipedia.org/wiki/Name_Service_Switch) on a typical Linux system; that's different from a username that you'd find in your cluster's audit logs.

Author: What do you mean with "a local user using `getent`"? And why is it relevant?

Reviewer: I mean that you could run `getent passwd | grep kubelet` and I would expect to learn the UID for `kubelet` by doing that. Is that the case?

This is because /etc/subuid and /etc/subgid map subordinate ranges to users, and if there isn't a user, what are we mapping it to?

Author: Yes, that is the case. However, those are not local users. That command includes local users and, for example, LDAP users.

>  * The binary `getsubids` installed (part of the [shadow-utils][shadow-utils] package)

Reviewer: We don't know what the package name is, because we don't know what OS they are using.

 * A configuration of subordinate UIDs/GIDs for the `kubelet` user (see `man 5
   subuid` and `man 5 subgid`)

> The `kubelet` user will only be used to gather the UID/GIDs range configuration,
> but the user the kubelet is executed as won't be changed by this.

Reviewer: ```suggestion
The kubelet looks for a user ID and group ID range assigned to the local user named `kubelet`;
the kubelet never switches its own real or effective group ID to that user.
```

Reviewer: > It is not a local user

If I run `getent passwd | grep kubelet`, do I see an entry? That's the test I'm imagining.

Reviewer: > I wouldn't say the kubelet will never switch the user. Other setups, like running the kubelet inside a userns, might at some point do that.

OK, in that case, we need a _different_ wording as 
> The `kubelet` user will only be used to gather the UID/GIDs range configuration, but the user the kubelet is executed as won't be changed by this.

might mislead people. What really happens?

Reviewer: > I wouldn't say the kubelet will never switch the user. Other setups, like running the kubelet inside a userns, might at some point do that.

OK,  so in long form, what information do we want the reader to learn from this paragraph around user switching? I'm not confident I know.

Author: @sftim 

> If I run `getent passwd | grep kubelet`, do I see an entry? That's the test I'm imagining.

I expect that, yes. But the output of that command is not just local users AFAIK.

> > The `kubelet` user will only be used to gather the UID/GIDs range configuration, but the user the kubelet is executed as won't be changed by this.
> 
> might mislead people. What really happens?

What happens is we check the mappings for the `kubelet`user and use that for the ranges. We don't change anything about the user used to run the kubelet.

> OK, so in long form, what information do we want the reader to learn from this paragraph around user switching? I'm not confident I know.

Exactly what I said in my previous answer.

Reviewer 2: Would this be more concise, but still accurate?

```suggestion
This setting only gathers the UID/GID range configuration and does not change the user executing the `kubelet`.
```

Author: Applied that, thanks.

> There are some requirements to configure the subordinate IDs:

Reviewer: ```suggestion
You must follow some constraints for the subordinate ID ranges that you assign to the local
`kubelet` user:
```

Author: Fixed, removed local user.

>  * The subordinate user ID must be a multiple of 65536 and greater than or equal
>    to it too. In other words, the first range 0-65535 can't be used for pods, as
>    that will circumvent some security benefits of user namespaces.

Reviewer: ```suggestion
* The subordinate user ID, that starts the UID range for Pods, **must** be a multiple of 65536
  and must also be greater than or equal to 65536. In other words, you cannot use any ID from the
  range 0-65535 for Pods; the kubelet imposes this restriction to make it difficult to create an
  accidentally insecure configuration.
```

Author: Applied

 * The subordinate ID count must be a multiple of 65536

>  * The subordinate ID count must be at least `65536 * <maxPods>` where `<maxPods>`

Reviewer: (nit)
```suggestion
* The subordinate ID count must be at least `65536 × <maxPods>` where `<maxPods>`
```

Author: Removed the initial space in all the lines of the list and applied this

   is the maximum number of pods that can run on the node.

>  * The configuration for UIDs and GIDs for the `kubelet` user must be identical

Reviewer: How about:
```suggestion
* For the `kubelet` user, you must assign the same range for both user IDs and for group IDs.
  It doesn't matter if other users have user ID ranges that don't align with the group ID ranges.

* None of the assigned ranges should overlap with any other assignment.
```

Author: At the beginning we already say this is for the `kubelet` user. I don't see why anyone would think this applies to any other user.

Also, it can overlap with other users if they want to. That is not _required_, we recommend it doesn't in another part already. But adding it here is mixing things, the kubelet initialization will fail if the things listed here are not met, that thing you added won't make it fail at all. And that is the expected behavior I want.

Author: I've applied that, but removed `for the kubelet user`. The text before the list says `You must follow some constraints for the subordinate ID range that you assign to the "kubelet" user` and all points apply only for the kubelet user.

Reviewer: > all points apply only for the kubelet user.

Although that's true, in docs we use repetition to help people take things in. Anyway, no need to redo this.

> For example, this can be set in `/etc/subuid` and `/etc/subgid`:

Reviewer: ```suggestion
For example, you could define `/etc/subuid` and `/etc/subgid` to both have contents
similar to:
```

Author: Added that. But changed the wording to:

```
For example, you could define `/etc/subuid` and `/etc/subgid` to both have
thsese entries for the `kubelet` user:
```

Before we were saying you could set that, but you were saying the file can be like this. Let's say what we were saying, the kubelet entry can look like this (we don't care about the rest of the file).

> ```shell

Reviewer: It's not a shell script.

Author: But I want shell highlighting...

Reviewer: Please don't mark this as a shell script. Marking things that aren't shell as shell can give us maintenance problems in the future, and so we ask contributors not to do that.

Author: Fixed.

# The format is

> #   user:firstID:count of IDs

Reviewer: ```suggestion
#   name:firstID:count of IDs
```
(Groups don't have usernames)

Author: Fixed

# where
# - firstID is 65536 (the minimum value possible)

> # - count of IDs is 110 (default max pods) * 65536

Reviewer: ```suggestion
# - count of IDs is 110 (default limit for number of pods) * 65536
```

Author: Applied

## Set up a node to support user namespaces

By default, the kubelet assigns pods UIDs/GIDs above the range 0-65535, based on
the assumption that the host's files and processes use UIDs/GIDs within this
range, which is standard for most Linux distributions. This approach prevents
any overlap between the UIDs/GIDs of the host and those of the pods.

Avoiding the overlap is important to mitigate the impact of vulnerabilities such
as [CVE-2021-25741][CVE-2021-25741], where a pod can potentially read arbitrary
files in the host. If the UIDs/GIDs of the pod and the host don't overlap, it is
limited what a pod would be able to do: the pod UID/GID won't match the host's
file owner/group.

The kubelet can use a custom range for user IDs and group IDs for pods. To
configure a custom range, the node needs to have:

 * A user `kubelet` in the system (you cannot use any other username here)
 * The binary `getsubids` installed (part of [shadow-utils][shadow-utils]) and
   in the `PATH` for the kubelet binary.

>  * A configuration of subordinate UIDs/GIDs for the `kubelet` user (see `[man 5
>    subuid](https://man7.org/linux/man-pages/man5/subuid.5.html)` and `[man 5
>    subgid](https://man7.org/linux/man-pages/man5/subgid.5.html)`)

Reviewer: This isn't rendering right. Try:
```suggestion
 * A configuration of subordinate UIDs/GIDs for the `kubelet` user (see
   [`man 5 subuid`](https://man7.org/linux/man-pages/man5/subuid.5.html) and
   [`man 5 subgid`](https://man7.org/linux/man-pages/man5/subgid.5.html)).
```

Author: Applied, thanks

## Set up a node to support user namespaces

> By default, the kubelet assumes host's files and host's processes use UIDs/GIDs
> in the range of 0-65535. This is the case in most Linux distributions with the
> default installation. The kubelet will assign UIDs/GIDs higher than that to
> pods, so there is no overlap between UID/GIDs used by host's files and processes
> and the ones used by pods.

Reviewer 2: ```suggestion
By default, the kubelet assigns pods UIDs/GIDs above the range 0-65535, based on the assumption that the host's files and processes use UIDs/GIDs within this range, which is standard for most Linux distributions. This approach prevents any overlap between the UIDs/GIDs of the host and those of the pods.
```

Author: Applied, thanks.

Avoiding the overlap is important to mitigate the impact of vulnerabilities such
as [CVE-2021-25741][CVE-2021-25741], where a pod can potentially read arbitrary
files in the host. If the UIDs/GIDs of the pod and the host don't overlap, it is
limited what a pod would be able to do: the pod UID/GID won't match the host's
file owner/group.

The kubelet can use a custom range for user IDs and group IDs for pods. To
configure a custom range, the node needs to have:

 * A user `kubelet` in the system (you cannot use any other username here)
 * The binary `getsubids` installed (part of [shadow-utils][shadow-utils]) and
   in the `PATH` for the kubelet binary.

>  * A configuration of subordinate UIDs/GIDs for the `kubelet` user (see `man 5
>    subuid` and `man 5 subgid`)

Reviewer 3: ```suggestion
 * A configuration of subordinate UIDs/GIDs for the `kubelet` user (see `[man 5
   subuid](https://man7.org/linux/man-pages/man5/subuid.5.html)` and `[man 5 subgid](https://man7.org/linux/man-pages/man5/subgid.5.html)`)
```

Author: Oh, great ideas, yes! I forget the manpages are available online :D

Author: Fixed, thanks!
